packages/libsql/src/lib/Category.ts

```diff
@@ -0,0 +1,417 @@
+import type {
+  IEventData,
+  StreamToken,
+  SyncResult,
+  ITimelineEvent,
+  TokenAndState,
+  ICodec,
+} from "@equinox-js/core"
+import * as Equinox from "@equinox-js/core"
+import * as Token from "./Token.js"
+import * as Read from "./Read.js"
+import { trace } from "@opentelemetry/api"
+import { Format, LibSqlReader, LibSqlWriter } from "./LibSqlClient.js"
+import { CachingCategory, ICachingStrategy, IReloadableCategory, Tags } from "@equinox-js/core"
+import { Client, Transaction } from "@libsql/client"
+import { randomUUID } from "crypto"
+
+const keepMap = Equinox.Internal.keepMap
+
+type GatewaySyncResult = { type: "Written"; token: StreamToken } | { type: "ConflictUnknown" }
+
+type Decode<E> = (v: ITimelineEvent<Format>) => E | undefined
+
+export class LibSqlConnection {
+  constructor(
+    public read: LibSqlReader,
+    public write: LibSqlWriter,
+  ) {}
+
+  static create(writeClient: Client, readClient = writeClient) {
+    return new LibSqlConnection(new LibSqlReader(readClient), new LibSqlWriter(writeClient))
+  }
+}
+
+type ContextConfig = {
+  client: Client
+  readClient?: Client
+  batchSize: number
+  maxBatches?: number
+}
+
+export class LibSqlContext {
+  constructor(
+    private readonly conn: LibSqlConnection,
+    public readonly batchSize: number,
+    public readonly maxBatches?: number,
+  ) {}
+
+  tokenEmpty = Token.create(0n)
+
+  async loadBatched<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    let state = initial
+    let version = 0n
+    let batches = 0
+    let eventCount = 0
+    for await (const [lastVersion, events] of Read.loadForwardsFrom(
+      this.conn.read,
+      this.batchSize,
+      this.maxBatches,
+      streamName,
+      0n,
+    )) {
+      batches++
+      eventCount += events.length
+      state = fold(state, keepMap(events, decode))
+      version = lastVersion
+    }
+    trace.getActiveSpan()?.setAttributes({
+      [Tags.loaded_count]: eventCount,
+      [Tags.batches]: batches,
+      [Tags.read_version]: Number(version),
+    })
+    return [Token.create(version), state]
+  }
+
+  async loadLast<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    const [version, events] = await Read.loadLastEvent(this.conn.read, streamName)
+    trace.getActiveSpan()?.setAttributes({
+      [Tags.loaded_count]: 1,
+      [Tags.read_version]: Number(version),
+    })
+    return [Token.create(version), fold(initial, keepMap(events, decode))]
+  }
+
+  async loadState<Event, State>(
```

Comment:
State doesnt convey much - either name it to convey somethign. about the stategy or add a comment to distinguish it from loadSnapshot (and maybe name that to convey that its loading for snapshot strategy more than just only literally loading a snapshot)